src/mono/mono/mini/intrinsics.c

```diff
@@ -488,6 +488,122 @@ emit_unsafe_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignatu
 		EMIT_NEW_BIALU (cfg, ins, OP_COMPARE, -1, args [0]->dreg, args [1]->dreg);
 		EMIT_NEW_UNALU (cfg, ins, OP_PCEQ, dreg, -1);
 		return ins;
+	} else if (!strcmp (cmethod->name, "BitCast")) {
+		g_assert (ctx);
+		g_assert (ctx->method_inst);
+		g_assert (ctx->method_inst->type_argc == 2);
+		g_assert (fsig->param_count == 1);
+
+		// We explicitly do not handle gsharedvt as it is meant as a slow fallback strategy
+		// instead we fallback to the managed implementation which will do the right things
+
+		MonoType *tfrom = ctx->method_inst->type_argv [0];
+		if (mini_is_gsharedvt_variable_type (tfrom)) {
+			return NULL;
+		}
+
+		MonoType *tto = ctx->method_inst->type_argv [1];
+		if (mini_is_gsharedvt_variable_type (tto)) {
+			return NULL;
+		}
+
+		// The underlying API always throws for reference type inputs, so we
+		// fallback to the managed implementation to let that handling occur
+
+		MonoTypeEnum tfrom_type = tfrom->type;
+		if (MONO_TYPE_IS_REFERENCE (tfrom)) {
```

Comment:
@lambdageek looks like the PR is almost there (🎉)

The remaining failure looks to be because this doesn't cover `nullable<t>` which should also be "blocked" (exit here so the managed impl is used instead).

Do you know if there's an easy way to detect that scenario?

Reply:
Looks like ther emight be `mono_class_is_nullable`, which requires checking the `klass`?

Reply:
Ok, that worked. There's two remaining issue and I'll definitely need some help on these ones....

-----------------

The first is that we have some failures like the following from: https://github.com/dotnet/runtime/blob/main/src/tests/JIT/Intrinsics/BitCast.il
```
Failed at line 18, expected 255, got 4294967295
Failed at line 19, expected -1, got 255
Failed at line 23, expected 65535, got 4294967295
Failed at line 24, expected -1, got 65535
```

These appear to be cases like `Bitcast<byte, byte>` where Mono isn't correctly handling the small types. Given the IL, I don't see any handling that I'd expect Mono to need to insert here, but its not getting it correct as is.

----------------

The other appears to be:
```
[08:58:58] info: [FAIL] System.Tests.HalfTests.ExplicitConversion_FromSingle(f: NaN, expected: NaN)
[08:58:58] info: Assert.Equal() Failure: Values differ
[08:58:58] info: Expected:   NaN
[08:58:58] info: Actual:     NaN
[08:58:58] info:    at System.Object.InvokeStub_HalfTests.ExplicitConversion_FromSingle(Object , Span`1 )
[08:58:58] info:    at System.Reflection.MethodBaseInvoker.InvokeWithFewArgs(Object , BindingFlags , Binder , Object[] , CultureInfo )
```

Where it looks like Mono might be normalizing NaN in a case where it shouldn't or doesn't need to. -- This is one of those cases where "technically" normalizing NaN is legal according to the IEEE 754 spec, but we ideally don't do that unnecessarily, particularly when it applies to `BitCast` or other constant folding operations.

I do see a comment in `mini-llvm` where it would have done such canonicalization, but it is currently under `#if 0` and is incorrectly stating that the WASM spec requires canonicalization when that is not the case. The WASM spec explicitly differentiates `canonical NaN` from `arithmetic NaN` and that the results of operations on NaN may be non-deterministic and even describes the standard callouts with regard to NaN propagation and its behavior for fneg, fabs, fcopysign, and fbits. It even describes an exact way to support defining `nan` with non-canonical payloads.

Given that it's commented out, it is unclear where this normalization might be occurring instead (perhaps LLVM itself is doing it?). -- I can always comment out the NaN tests for WASM, but it'd be more ideal if we didn't have to.

Reply:
The WASM spec allows implementations to do weird things to your NaNs, so asserting that they are bitwise equal will probably cause test flakiness for WASM lanes. In some cases it explicitly says it will do weird things to them.